browser/brave_vpn/dns/brave_vpn_dns_observer_service_unittest.cc

```diff
@@ -0,0 +1,317 @@
+    bool callback_called = false;
+    dns_service_->SetVPNNotificationCallbackForTesting(
+        base::BindLambdaForTesting([&]() { callback_called = true; }));
+    SetDNSMode(mode, servers);
+    return callback_called;
+  }
+
+  void SetDNSMode(const std::string& mode, const std::string& doh_providers) {
+    local_state()->SetString(::prefs::kDnsOverHttpsTemplates, doh_providers);
+    local_state()->SetString(::prefs::kDnsOverHttpsMode, mode);
+    SystemNetworkContextManager::GetStubResolverConfigReader()
+        ->UpdateNetworkService(false);
+  }
+
+  bool WasPolicyNotificationShownForState(mojom::ConnectionState state) {
+    bool callback_called = false;
+    dns_service_->SetPolicyNotificationCallbackForTesting(
+        base::BindLambdaForTesting([&]() { callback_called = true; }));
+    FireBraveVPNStateChange(state);
+    return callback_called;
+  }
+
+  void ExpectDNSMode(const std::string& mode,
+                     const std::string& doh_providers) {
+    auto dns_config = SystemNetworkContextManager::GetStubResolverConfigReader()
+                          ->GetSecureDnsConfiguration(false);
+    auto* current_mode = SecureDnsConfig::ModeToString(dns_config.mode());
+    auto current_servers = dns_config.doh_servers().ToString();
+    EXPECT_EQ(current_mode, mode);
+    EXPECT_EQ(current_servers, doh_providers);
+  }
+  void SetManagedMode(const std::string& value) {
+    local_state_.SetManagedPref(::prefs::kDnsOverHttpsMode,
+                                std::make_unique<base::Value>(value));
+  }
+
+ private:
+  std::unordered_map<std::string, std::string> policy_map_;
+  content::BrowserTaskEnvironment task_environment_;
+  std::unique_ptr<BraveVpnDnsObserverService> dns_service_;
+  sync_preferences::TestingPrefServiceSyncable profile_pref_service_;
+  TestingPrefServiceSimple local_state_;
+  std::unique_ptr<StubResolverConfigReader> stub_resolver_config_reader_;
+};
+
+TEST_F(BraveVpnDnsObserverServiceUnitTest, AutoEnable) {
+  // Browser DoH mode off -> override browser config and enable vpn
+  local_state()->ClearPref(::prefs::kBraveVpnDnsConfig);
+  SetDNSMode(SecureDnsConfig::kModeOff, "");
+  FireBraveVPNStateChange(mojom::ConnectionState::CONNECTING);
+  ExpectDNSMode(SecureDnsConfig::kModeOff, "");
+  FireBraveVPNStateChange(mojom::ConnectionState::CONNECTED);
+  ExpectDNSMode(SecureDnsConfig::kModeSecure, kCloudflareDnsProviderURL);
+  EXPECT_FALSE(local_state()->GetString(::prefs::kBraveVpnDnsConfig).empty());
+  FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTING);
+  ExpectDNSMode(SecureDnsConfig::kModeSecure, kCloudflareDnsProviderURL);
+  FireBraveVPNStateChange(mojom::ConnectionState::DISCONNECTED);
+  ExpectDNSMode(SecureDnsConfig::kModeOff, "");
+  EXPECT_TRUE(local_state()->GetString(::prefs::kBraveVpnDnsConfig).empty());
+
+  // Browser DoH mode automatic -> override browser config and enable vpn
+  SetDNSMode(SecureDnsConfig::kModeAutomatic, "");
```

Comment:
Please split these up into smaller test cases. Long test cases make it harder to verify that all cases are covered and  it's also harder to figure out which specific scenario failed